app/controllers/invites_controller.rb

```diff
@@ -23,84 +20,56 @@ def show
-          invite.redeem_invite_link(email: params[:email], username: params[:username], name: params[:name], password: params[:password], user_custom_fields: params[:user_custom_fields], ip_address: request.remote_ip)
-        else
-          invite.redeem(username: params[:username], name: params[:name], password: params[:password], user_custom_fields: params[:user_custom_fields], ip_address: request.remote_ip)
-        end
-
-        if user.present?
-          log_on_user(user) if user.active?
-          user.update_timezone_if_missing(params[:timezone])
-          post_process_invite(user)
-          response = { success: true }
-        else
-          response = { success: false, message: I18n.t('invite.not_found_json') }
-        end
-
-        if user.present? && user.active?
-          topic = invite.topics.first
-          response[:redirect_to] = topic.present? ? path("#{topic.relative_url}") : path("/")
-        elsif user.present?
-          response[:message] = I18n.t('invite.confirm_email')
-        end
+  def create
+    guardian.ensure_can_send_invite_links!(current_user) if params[:email].blank? || params[:skip_email].present?
 
-        render json: response
-      rescue ActiveRecord::RecordInvalid, ActiveRecord::RecordNotSaved => e
-        render json: {
-          success: false,
-          errors: e.record&.errors&.to_hash || {},
-          message: I18n.t('invite.error_message')
-        }
-      rescue Invite::UserExists => e
-        render json: { success: false, message: [e.message] }
-      end
-    else
-      render json: { success: false, message: I18n.t('invite.not_found_json') }
+    if params[:email].present? && Invite.exists?(email: params[:email])
+      return render json: failed_json, status: 422
     end
-  end
 
-  def create
-    params.require(:email)
+    if params[:topic_id].present?
+      topic = Topic.find_by(id: params[:topic_id])
+      guardian.ensure_can_invite_to!(topic)
+    end
 
-    groups = Group.lookup_groups(
-      group_ids: params[:group_ids],
-      group_names: params[:group_names]
-    )
+    if params[:group_ids].present? || params[:group_names].present?
+      groups = Group.lookup_groups(group_ids: params[:group_ids], group_names: params[:group_names])
+    end
 
     guardian.ensure_can_invite_to_forum!(groups)
-    group_ids = groups.map(&:id)
-
-    if Invite.exists?(email: params[:email])
-      return render json: failed_json, status: 422
-    end
 
     begin
-      if Invite.invite_by_email(params[:email], current_user, nil, group_ids, params[:custom_message])
-        render json: success_json
+      invite = Invite.generate(current_user,
+        invite_key: params[:invite_key],
+        email: params[:email],
+        skip_email: params[:skip_email],
+        invited_by: current_user,
+        custom_message: params[:custom_message],
+        max_redemptions_allowed: params[:max_redemptions_allowed],
+        topic_id: topic&.id,
+        group_ids: groups,
```

Comment:
I think we need to pluck the ids here? `groups` is currently an ActiveRecord association and according to the keyword param, it is expecting ids.